File: src/types/node.ts
/** Indexed Node.js package metadata returned by `node_list_projects`. */
export interface NodeProjectNode {
  projectId: number
  repositoryId: number | null
  path: string
  name: string
  version: string
  packageManager: string | null
  /** JSON object text preserving package.json script order. */
  scriptsJson: string
  /** workspace 根目录（N-09）；独立工程为 null。 */
  workspaceRoot: string | null
}

// Union declarations end with `;` — the IPC golden parser (models/ipc_golden)
// terminates a union body on a `;`-ended line and skips pure string-literal
// unions, so these must not rely on ASI.
export type NodeExecutableKind = "node" | "packageManager";
export type NodePackageManager = "npm" | "pnpm" | "yarn" | "bun";

export interface NodeExecutable {
  id: number | null
  kind: NodeExecutableKind
  packageManager: NodePackageManager | null
  executablePath: string
  version: string | null
  rawOutput: string
  isValid: boolean
  lastChecked: string
  createdAt: string | null
  updatedAt: string | null
}

export interface NodeExecutableRequest {
  kind: NodeExecutableKind
  packageManager?: NodePackageManager | null
  executablePath: string
}

export interface NodeInstallRequest {
  projectDir: string
  packageManager: NodePackageManager
  confirmed?: boolean
}

/** N-10：本机扫描出的工具链候选（`node_scan_executables` 返回，只读发现）。 */
export interface NodeScanCandidate {
  kind: NodeExecutableKind
  packageManager: NodePackageManager | null
  executablePath: string
  /** `-v` 探测版本；探测失败为 null（「未知版本」）。 */
  version: string | null
  probeOk: boolean
  /** 扫描来源标签：system / nvm / nvm-windows / fnm / volta / mise / asdf / n / scoop / homebrew / pnpm-home / yarn-home。 */
  source: string
  /** 该路径是否已在注册表中（禁选防重复登记）。 */
  registered: boolean
}

File: src-tauri/src/models/ipc_golden/node.rs
//! Node.js IPC samples (N-02 / N-08).

use std::path::PathBuf;

use serde_json::{json, Map, Value};

use crate::commands::node::NodeInstallRequest;
use crate::node::model::{NodeExecutable, NodeExecutableKind, NodeExecutableRequest, PackageManager};
use crate::node::NodeProjectNode;

pub(super) fn samples(m: &mut Map<String, Value>) {
    m.insert(
        "NodeProjectNode".into(),
        json!(NodeProjectNode {
            project_id: 7,
            repository_id: Some(3),
            path: PathBuf::from("/ws/web"),
            name: "web".into(),
            version: "1.2.3".into(),
            package_manager: Some("npm".into()),
            scripts_json: r#"{"dev":"vite","build":"vite build"}"#.into(),
            pkg_hash: "0123456789abcdef".into(),
        }),
    );
    m.insert(
        "NodeExecutable".into(),
        json!(NodeExecutable {
            id: Some(1),
            kind: NodeExecutableKind::PackageManager,
            package_manager: Some(PackageManager::Pnpm),
            executable_path: "/usr/local/bin/pnpm".into(),
            version: Some("11.24.0".into()),
            raw_output: "11.24.0".into(),
            is_valid: true,
            last_checked: "2026-09-01T00:00:00+00:00".into(),
            created_at: Some("2026-09-01T00:00:00+00:00".into()),
            updated_at: Some("2026-09-01T00:00:00+00:00".into()),
        }),
    );
    m.insert(
        "NodeExecutableRequest".into(),
        json!(NodeExecutableRequest {
            kind: NodeExecutableKind::PackageManager,
            package_manager: Some(PackageManager::Pnpm),
            executable_path: "/usr/local/bin/pnpm".into(),
        }),
    );
    m.insert(
        "NodeInstallRequest".into(),
        json!(NodeInstallRequest {
            project_dir: "/ws/web".into(),
            package_manager: PackageManager::Pnpm,
            confirmed: false,
        }),
    );
}

pub(super) const TS_TYPE_MAP: &[(&str, &str, &str)] = &[
    ("NodeProjectNode", "types/node.ts", "NodeProjectNode"),
    ("NodeExecutable", "types/node.ts", "NodeExecutable"),
    ("NodeExecutableRequest", "types/node.ts", "NodeExecutableRequest"),
    ("NodeInstallRequest", "types/node.ts", "NodeInstallRequest"),
];
